x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts

```diff
@@ -0,0 +1,1006 @@
+        mount({
+          apiUrl: '',
+        });
+        CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+        expect(CredentialsLogic.values.apiUrl).toEqual('http://example.com');
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setCredentialsDetails(credentialsDetails);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        engines: expect.any(Array),
+        apiUrl: expect.any(String),
+        isCredentialsDetailsComplete: expect.any(Boolean),
+      });
+    });
+  });
+
+  describe('setNameInputBlurred', () => {
+    describe('nameInputBlurred', () => {
+      it('should set this value', () => {
+        mount({
+          nameInputBlurred: false,
+        });
+        CredentialsLogic.actions.setNameInputBlurred(true);
+        expect(CredentialsLogic.values.nameInputBlurred).toEqual(true);
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setNameInputBlurred(true);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        nameInputBlurred: expect.any(Boolean),
+      });
+    });
+  });
+
+  describe('setTokenReadWrite', () => {
+    describe('activeApiToken', () => {
+      // TODO should probably have a more specific behavior
+      it('sets an arbitrary key and value on the activeApiToken', () => {
+        mount({
+          activeApiToken: {
+            ...newToken,
+            foo: 'bar',
+          },
+        });
+        CredentialsLogic.actions.setTokenReadWrite({ name: 'foo', checked: 'whatever' });
+        expect(CredentialsLogic.values.activeApiToken).toEqual({
+          ...newToken,
+          foo: 'whatever',
+        });
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setTokenReadWrite({ name: 'foo', checked: 'whatever' });
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        activeApiToken: expect.any(Object),
+      });
+    });
+  });
+
+  describe('setTokenName', () => {
+    describe('activeApiToken', () => {
+      it('update the name property on the activeApiToken, formatted correctly', () => {
+        mount({
+          activeApiToken: {
+            ...newToken,
+            name: 'bar',
+          },
+        });
+        CredentialsLogic.actions.setTokenName('New Name');
+        expect(CredentialsLogic.values.activeApiToken).toEqual({
+          ...newToken,
+          name: 'new-name',
+        });
+      });
+    });
+
+    describe('activeApiTokenRawName', () => {
+      it('update the raw name, with no formatting applied', () => {
+        mount();
+        CredentialsLogic.actions.setTokenName('New Name');
+        expect(CredentialsLogic.values.activeApiTokenRawName).toEqual('New Name');
+      });
+    });
+
+    describe('keyHelpText', () => {
```

Comment:
This is actually a selector, not a value. Selectors are tested separately. However, to a consumer of the logic that is transparent. To the consumer of the logic, they call the `activeApiTokenRawName` and the `keyHelpText` value is updated. So since this is part of the public API of the `activeApiTokenRawName` action I added it here. I also add more thorough tests to `keyHelpText` below.

Reply:
FWIW I'm ++ for testing selectors in the exact same way that we test reducers/all values.

Reply:
Although I have to say... is `keyHelpText` really something we need to be be storing in our logic file? I'm not really following why it can't simply be a simple React conditional, e.g.:

```js
const { activeApiToken, activeApiTokenRawName } = useValues(CredentialsLogic);
const hasFormattedName = activeApiToken.name && activeApiToken.name !== activeApiTokenRawName;

return (
  <>
    {hasFormattedName && (
      <div className="someHelpText">
        Your key will be named: {activeApiToken.name}
      </div>
    )}
  </>
);
```

I mention this because we need to `i18n.translate()` "Your key will be named" - and while we can still definitely do so in the logic file, it's somewhat easier to put all your user-facing strings etc. in the same place 🤷

Reply:
Ah, that's a good point.

Reply:
I removed keyHelpText, will figure out a front-end solution when we get there: 3e1fe7d